test/e2e/subroutes_test.go

```diff
@@ -85,3 +91,82 @@ func TestSubrouteLocalSTS(t *testing.T) { // We can't use a longer more descript
 		})
 	}
 }
+
+func TestSubrouteVisibilityChange(t *testing.T) {
+	t.Parallel()
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	clients := Setup(t)
+
+	t.Log("Creating a Service for the helloworld test app.")
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "helloworld",
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	tag := "my-tag"
+
+	withInternalVisibility := WithServiceLabel(
+		routeconfig.VisibilityLabelKey, routeconfig.VisibilityClusterLocal)
+	withTrafficSpec := WithInlineRouteSpec(v1alpha1.RouteSpec{
+		Traffic: []v1alpha1.TrafficTarget{
+			{
+				TrafficTarget: v1beta1.TrafficTarget{
+					Tag:            tag,
+					Percent:        100,
+				},
+			},
+		},
+	})
+	resources, err := v1a1test.CreateRunLatestServiceReady(t, clients, &names, &v1a1test.Options{}, withInternalVisibility, withTrafficSpec)
+	if err != nil {
+		t.Fatalf("Failed to create initial Service: %v: %v", names.Service, err)
+	}
+
+	// Ensure that it only has cluster local addresses
+	for _, traffic := range resources.Route.Status.Traffic {
+		if !strings.HasSuffix(traffic.TrafficTarget.URL.Host, network.GetClusterDomainName()) {
+			t.Fatalf("Expected all subroutes to be cluster local")
+		}
+	}
+
+	seriviceName := fmt.Sprintf("%s-%s", tag, resources.Route.Name)
+	svc, err := clients.KubeClient.Kube.CoreV1().Services(test.ServingNamespace).Get(seriviceName, v1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Failed to get k8s service to modify: %s", err.Error())
+	}
+
+	svcCopy := svc.DeepCopy()
+	labels.SetVisibility(&svcCopy.ObjectMeta, false)
+
+	patch, err := duck.CreatePatch(svc, svcCopy)
+	if err != nil {
+		t.Fatalf("Failed to create patch: %s", err.Error())
+	}
+
+	encPatch, err := patch.MarshalJSON()
+	if err != nil {
+		t.Fatalf("Failed to marshal patch: %s", err.Error())
+	}
+	_, err = clients.KubeClient.Kube.CoreV1().Services(test.ServingNamespace).Patch(seriviceName, types.JSONPatchType, encPatch)
+	if err != nil {
+		t.Fatalf("Failed to patch service: %s", err.Error())
+	}
+
+	// Get updated route to check.
+	time.Sleep(5 * time.Second) // Wait for changes to propagate..
```

Comment:
Is it possible to create the placeholder Service with the visibility label even before Route?   That way you don't have to do the wait, which makes the test runs longer and flakier.

Alternatively, you can probably use ObservedGeneration in the Route Status.  However currently we have incorrect `ClusterIngress#Status` since the child VirtualService doesn't provide any, I don't think this path would work very well.

Reply:
I don't like the sleep either, I was hoping there was other known ways of checking for that. Will dig around.

Creating the service beforehand won't work since we ensure that we fail early if the service with the same name as one created by the route exists.

Reply:
Is it possible to create something similar to [`WaitForServiceState`](https://github.com/knative/serving/blob/master/test/v1alpha1/service.go#L307) but for routes?

Reply:
Looks like there is a `WaitForRouteState` already, thanks for pointing that out @taragu!